File: sdk/typescript/src/agent/processLogs.ts
type LogEntry = {
  v: number;
  seq: number;
  ts: string;
  stream: string;
  line: string;
  level?: string;
  source?: string;
  truncated?: boolean;
};

function logsEnabled(): boolean {
  const v = (process.env.AGENTFIELD_LOGS_ENABLED ?? 'true').trim().toLowerCase();
  return !['0', 'false', 'no', 'off'].includes(v);
}

function maxBufferBytes(): number {
  const raw = parseInt(process.env.AGENTFIELD_LOG_BUFFER_BYTES ?? '4194304', 10);
  return Number.isFinite(raw) && raw >= 1024 ? raw : 4194304;
}

function maxLineBytes(): number {
  const raw = parseInt(process.env.AGENTFIELD_LOG_MAX_LINE_BYTES ?? '16384', 10);
  return Number.isFinite(raw) && raw >= 256 ? raw : 16384;
}

function maxTailLines(): number {
  const raw = parseInt(process.env.AGENTFIELD_LOG_MAX_TAIL_LINES ?? '50000', 10);
  return Number.isFinite(raw) && raw >= 1 ? raw : 50000;
}

function internalBearerOk(authHeader: string | undefined): boolean {
  const want = (process.env.AGENTFIELD_AUTHORIZATION_INTERNAL_TOKEN ?? '').trim();
  if (!want) return true;
  if (!authHeader?.toLowerCase().startsWith('bearer ')) return false;
  return authHeader.slice(7).trim() === want;
}

export class ProcessLogRing {
  private seq = 0;
  private entries: LogEntry[] = [];
  private approxBytes = 0;
  private readonly maxBytes: number;

  constructor() {
    this.maxBytes = maxBufferBytes();
  }

  append(stream: string, line: string, truncated: boolean): void {
    const ts = new Date().toISOString();
    this.seq += 1;
    const sl = stream.toLowerCase();
    const level = sl === 'stderr' ? 'error' : sl === 'stdout' ? 'info' : 'log';
    const e: LogEntry = {
      v: 1,
      seq: this.seq,
      ts,
      stream,
      line,
      level,
      source: 'process',
      truncated,
    };
    this.entries.push(e);
    this.approxBytes += line.length + 64;
    while (this.approxBytes > this.maxBytes && this.entries.length > 1) {
      const old = this.entries.shift()!;
      this.approxBytes -= old.line.length + 64;
    }
  }

  tail(n: number): LogEntry[] {
    if (n <= 0) return [];
    return this.entries.length <= n ? [...this.entries] : this.entries.slice(-n);
  }

  snapshotAfter(sinceSeq: number, limit: number | null): LogEntry[] {
    const buf = this.entries.filter((e) => e.seq > sinceSeq);
    if (limit != null && limit > 0 && buf.length > limit) {
      return buf.slice(-limit);
    }
    return buf;
  }

}

let captureInstalled = false;

export function installStdioLogCapture(ring: ProcessLogRing): void {
  if (captureInstalled || !logsEnabled()) return;
  captureInstalled = true;
  const maxLB = maxLineBytes();

  const hook = (stream: NodeJS.WriteStream, name: string) => {
    const orig = stream.write.bind(stream);
    let buf = '';
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (stream as any).write = (chunk: any, ...args: any[]) => {
      const s =
        typeof chunk === 'string'
          ? chunk
          : Buffer.isBuffer(chunk)
            ? chunk.toString('utf8')
            : String(chunk);
      buf += s;
      let idx: number;
      while ((idx = buf.indexOf('\n')) >= 0) {
        const line = buf.slice(0, idx);
        buf = buf.slice(idx + 1);
        let out = line;
        let trunc = false;
        if (Buffer.byteLength(out, 'utf8') > maxLB) {
          out = Buffer.from(out, 'utf8').subarray(0, maxLB).toString('utf8');
          trunc = true;
        }
        ring.append(name, out, trunc);
      }
      return orig(chunk, ...args);
    };
  };

  hook(process.stdout, 'stdout');
  hook(process.stderr, 'stderr');
}

export function registerAgentfieldLogsRoute(
  app: import('express').Express,
  ring: ProcessLogRing
): void {
  app.get('/agentfield/v1/logs', (req, res) => {
    if (!logsEnabled()) {
      return res.status(404).json({
        error: 'logs_disabled',
        message: 'Process logs API is disabled'
      });
    }
    const auth = req.headers.authorization;
    if (!internalBearerOk(auth)) {
      return res.status(401).json({
        error: 'unauthorized',
        message: 'Valid Authorization Bearer required'
      });
    }
    let tailLines = parseInt(String(req.query.tail_lines ?? '0'), 10);
    const sinceSeq = parseInt(String(req.query.since_seq ?? '0'), 10);
    const follow = ['1', 'true', 'yes'].includes(String(req.query.follow ?? '').toLowerCase());
    const cap = maxTailLines();
    if (tailLines > cap) {
      return res.status(413).json({
        error: 'tail_too_large',
        message: `tail_lines exceeds max ${cap}`
      });
    }
    if (tailLines <= 0 && sinceSeq <= 0 && !follow) tailLines = 200;

    let initial: LogEntry[];
    if (sinceSeq > 0) {
      initial = ring.snapshotAfter(sinceSeq, tailLines > 0 ? tailLines : null);
    } else {
      const n = tailLines > 0 ? tailLines : 200;
      initial = ring.tail(n);
    }

    res.setHeader('Content-Type', 'application/x-ndjson');
    res.setHeader('Cache-Control', 'no-store');
    res.status(200);

    const writeLines = (entries: LogEntry[]) => {
      for (const e of entries) {
        res.write(`${JSON.stringify(e)}\n`);
      }
      (res as import('express').Response & { flush?: () => void }).flush?.();
    };

    writeLines(initial);
    let lastSeq = sinceSeq;
    if (initial.length) lastSeq = initial[initial.length - 1]!.seq;

    if (!follow) {
      return res.end();
    }

    const iv = setInterval(() => {
      const newer = ring.snapshotAfter(lastSeq, null);
      if (newer.length) {
        writeLines(newer);
        lastSeq = newer[newer.length - 1]!.seq;
      }
    }, 400);

    req.on('close', () => {
      clearInterval(iv);
    });
  });
}

File: sdk/go/agent/process_logs.go
package agent

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

// processLogEntry matches NDJSON v1 (docs/api/AGENT_NODE_LOGS.md).
type processLogEntry struct {
	V         int    `json:"v"`
	Seq       int    `json:"seq"`
	TS        string `json:"ts"`
	Stream    string `json:"stream"`
	Line      string `json:"line"`
	Truncated bool   `json:"truncated,omitempty"`
}

type processLogRing struct {
	mu          sync.Mutex
	seq         int
	entries     []processLogEntry
	approxBytes int
	maxBytes    int
	notify      sync.Cond
}

func newProcessLogRing(maxBytes int) *processLogRing {
	if maxBytes < 1024 {
		maxBytes = 1024
	}
	r := &processLogRing{maxBytes: maxBytes}
	r.notify.L = &r.mu
	return r
}

func processLogsMaxBytes() int {
	raw := strings.TrimSpace(os.Getenv("AGENTFIELD_LOG_BUFFER_BYTES"))
	if raw == "" {
		return 4 << 20
	}
	n, err := strconv.Atoi(raw)
	if err != nil || n < 1024 {
		return 4 << 20
	}
	return n
}

func processLogsMaxLineBytes() int {
	raw := strings.TrimSpace(os.Getenv("AGENTFIELD_LOG_MAX_LINE_BYTES"))
	if raw == "" {
		return 16384
	}
	n, err := strconv.Atoi(raw)
	if err != nil || n < 256 {
		return 16384
	}
	return n
}

func processLogsMaxTailLines() int {
	raw := strings.TrimSpace(os.Getenv("AGENTFIELD_LOG_MAX_TAIL_LINES"))
	if raw == "" {
		return 50000
	}
	n, err := strconv.Atoi(raw)
	if err != nil || n < 1 {
		return 50000
	}
	return n
}

func processLogsEnabled() bool {
	v := strings.ToLower(strings.TrimSpace(os.Getenv("AGENTFIELD_LOGS_ENABLED")))
	return v != "0" && v != "false" && v != "no" && v != "off"
}

func internalBearerOK(authHeader string) bool {
	want := strings.TrimSpace(os.Getenv("AGENTFIELD_AUTHORIZATION_INTERNAL_TOKEN"))
	if want == "" {
		return true
	}
	if !strings.HasPrefix(strings.ToLower(authHeader), "bearer ") {
		return false
	}
	got := strings.TrimSpace(authHeader[7:])
	return got == want
}

func (r *processLogRing) appendLine(stream, line string, truncated bool) {
	if r == nil {
		return
	}
	ts := time.Now().UTC().Format("2006-01-02T15:04:05.000Z07:00")
	r.mu.Lock()
	r.seq++
	e := processLogEntry{V: 1, Seq: r.seq, TS: ts, Stream: stream, Line: line, Truncated: truncated}
	r.entries = append(r.entries, e)
	r.approxBytes += len(line) + 64
	for r.approxBytes > r.maxBytes && len(r.entries) > 1 {
		old := r.entries[0]
		r.entries = r.entries[1:]
		r.approxBytes -= len(old.Line) + 64
	}
	r.notify.Broadcast()
	r.mu.Unlock()
}

func (r *processLogRing) tail(n int) []processLogEntry {
	r.mu.Lock()
	defer r.mu.Unlock()
	if n <= 0 {
		return nil
	}
	if len(r.entries) <= n {
		out := make([]processLogEntry, len(r.entries))
		copy(out, r.entries)
		return out
	}
	out := make([]processLogEntry, n)
	copy(out, r.entries[len(r.entries)-n:])
	return out
}

func (r *processLogRing) snapshotAfter(sinceSeq int, limit int) []processLogEntry {
	r.mu.Lock()
	defer r.mu.Unlock()
	var buf []processLogEntry
	for _, e := range r.entries {
		if e.Seq > sinceSeq {
			buf = append(buf, e)
		}
	}
	if limit > 0 && len(buf) > limit {
		buf = buf[len(buf)-limit:]
	}
	return buf
}

func (a *Agent) ensureProcessLogRing() {
	if a == nil {
		return
	}
	a.procLogOnce.Do(func() {
		if !processLogsEnabled() {
			return
		}
		a.procLogRing = newProcessLogRing(processLogsMaxBytes())
		installProcessStdioCapture(a.procLogRing)
	})
}

func (a *Agent) handleAgentfieldLogs(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}
	if !processLogsEnabled() {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusNotFound)
		_, _ = w.Write([]byte(`{"error":"logs_disabled","message":"Process logs API is disabled"}`))
		return
	}
	a.ensureProcessLogRing()
	if a.procLogRing == nil {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusNotFound)
		_, _ = w.Write([]byte(`{"error":"logs_disabled","message":"Process logs API is disabled"}`))
		return
	}
	if !internalBearerOK(r.Header.Get("Authorization")) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusUnauthorized)
		_, _ = w.Write([]byte(`{"error":"unauthorized","message":"Valid Authorization Bearer required"}`))
		return
	}
	q := r.URL.Query()
	tailLines, _ := strconv.Atoi(q.Get("tail_lines"))
	sinceSeq, _ := strconv.Atoi(q.Get("since_seq"))
	follow := strings.EqualFold(q.Get("follow"), "1") || strings.EqualFold(q.Get("follow"), "true") || q.Get("follow") == "yes"
	maxTail := processLogsMaxTailLines()
	if tailLines > maxTail {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusRequestEntityTooLarge)
		_, _ = fmt.Fprintf(w, `{"error":"tail_too_large","message":"tail_lines exceeds max %d"}`, maxTail)
		return
	}
	if tailLines <= 0 && sinceSeq <= 0 && !follow {
		tailLines = 200
	}

	var initial []processLogEntry
	if sinceSeq > 0 {
		initial = a.procLogRing.snapshotAfter(sinceSeq, tailLines)
	} else {
		n := tailLines
		if n <= 0 {
			n = 200
		}
		initial = a.procLogRing.tail(n)
	}

	w.Header().Set("Content-Type", "application/x-ndjson")
	w.Header().Set("Cache-Control", "no-store")
	w.WriteHeader(http.StatusOK)
	fl, _ := w.(http.Flusher)

	enc := json.NewEncoder(w)
	for _, e := range initial {
		if err := enc.Encode(e); err != nil {
			return
		}
		if fl != nil {
			fl.Flush()
		}
	}
	if !follow {
		return
	}

	lastSeq := sinceSeq
	if len(initial) > 0 {
		lastSeq = initial[len(initial)-1].Seq
	}

	tick := time.NewTicker(400 * time.Millisecond)
	defer tick.Stop()
	ctx := r.Context()

	for {
		select {
		case <-ctx.Done():
			return
		case <-tick.C:
			newer := a.procLogRing.snapshotAfter(lastSeq, 0)
			for _, e := range newer {
				if err := enc.Encode(e); err != nil {
					return
				}
				lastSeq = e.Seq
			}
			if fl != nil {
				fl.Flush()
			}
		}
	}
}
